internal/actions/reconcile_components_test.go

```diff
@@ -0,0 +1,245 @@
+package actions
+
+import (
+	"github.com/broadinstitute/revere/internal/configuration"
+	"github.com/broadinstitute/revere/internal/statuspage"
+	"github.com/go-resty/resty/v2"
+	"github.com/jarcoal/httpmock"
+	"reflect"
+	"sort"
+	"testing"
+)
+
+func TestReconcileComponents(t *testing.T) {
+	type args struct {
+		config *configuration.Config
+		client *resty.Client
+	}
+	config := configuration.Config{
+		Verbose: false,
+		Client: struct {
+			Redirects int
+			Retries   int
+		}{Redirects: 0, Retries: 0},
+		Statuspage: struct {
+			ApiKey     string `validate:"required"`
+			PageID     string `validate:"required"`
+			ApiRoot    string
+			Components []configuration.Component
+		}{ApiKey: "key", PageID: "foo", ApiRoot: "https://localhost",
+			Components: []configuration.Component{
+				{Name: "Same", Description: "Same description"},
+				{Name: "Modified", Description: "New description"},
+				{Name: "New", Description: "A new component"},
+			},
+		},
+	}
+	client := statuspage.Client(&config)
+	tests := []struct {
+		name string
+		args args
+		// ID-component maps to use for the mock
+		start   map[string]statuspage.Component
+		end     map[string]statuspage.Component
+		wantErr bool
+	}{
+		{
+			name: "Reconciles components",
+			args: args{
+				config: &config,
+				client: client,
+			},
+			start: map[string]statuspage.Component{
+				"1": {Name: "Same", Description: "Same description", Showcase: true, Status: "operational", ID: "1", PageID: "foo"},
+				"2": {Name: "Modified", Description: "Old description", Showcase: true, Status: "operational", ID: "2", PageID: "foo"},
+				"3": {Name: "Deleted", Description: "To be deleted", Showcase: true, Status: "operational", ID: "3", PageID: "foo"},
+			},
+			end: map[string]statuspage.Component{
+				"1": {Name: "Same", Description: "Same description", Showcase: true, Status: "operational", ID: "1", PageID: "foo"},
+				"2": {Name: "Modified", Description: "New description", Showcase: true, Status: "operational", ID: "2", PageID: "foo"},
+				"4": {Name: "New", Description: "A new component", Showcase: true, Status: "operational", ID: "4", PageID: "foo"},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			httpmock.ActivateNonDefault(tt.args.client.GetClient())
```

Comment:
does this autogenerate a mock of the status page api. Or is this a mock of resty?

Reply:
The mock of the API is in component_api_mock.go, we call the config function right below this. This line turns the mocking system on and tells it to intercept resty (tt.args.client is a resty.Client, .GetClient gives a raw net/http client). Both httpmock and resty specifically document this mechanism for integrating with each other, go figure.